var/spack/repos/builtin/packages/intel-oneapi/package.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import tempfile
+import os
+import glob
+
+
+class IntelOneapi(Package):
+    """
+    Includes the icx/ifx compiler executables.
+    """
+
+    homepage = "https://software.intel.com/content/www/us/en/develop/tools/oneapi.html"
+
+    version('2021.1.9.2205', sha256='8926a3001e61edbb293cb607beb8eb3a3511330a4625c8f3b1b51603426f121a', url='https://registrationcenter-download.intel.com/akdlm/irc_nas/16949/l_HPCKit_b_2021.1.9.2205_offline.sh', expand=False)
+
+    phases = ['install']
+
+    def install(self, spec, prefix):
+        bash = Executable('bash')
+
+        # Capture logs written in /tmp
+        tmpdir = tempfile.mkdtemp(prefix='spack-intel-')
+        bash.add_default_env('TMPDIR', tmpdir)
+
+        # Need to set HOME to avoid using ~/intel
+        bash.add_default_env('HOME', prefix)
+
+        bash('./l_HPCKit_b_%s_offline.sh' %
+             spec.versions.lowest(),
+             '-s', '-a', '-s', '--action', 'install',
+             '--eula', 'accept',
+             '--components',
+             ('intel.oneapi.lin.dpcpp-cpp-compiler-pro'
```

Comment:
This package will be hardwired to install the compilers. What do you think about having a base oneapi class, where components is a member that is overridden by derived classes InteloneAPICompilers?

Reply:
What are all the components that are available? Could it make sense to install them all here?

> What do you think about having a base oneapi class, where components is a member that is overridden by derived classes InteloneAPICompilers?

This could be useful. It would be simplest if the component sets did not overlap between packages. Do you see a reason for Spack packages with overlapping component sets (e.g. two packages which include `intel.oneapi.lin.dpcpp-cpp-compiler-pro` in `--components`)?

Reply:
compiler, onemkl, onedal, oneccl, onednn, onetbb. TBB is relatively small (< 100 mbytes) and DAL and MKL are much larger 500+ mbytes. Downloading and installing them all would not be convenient for some people.

There are some dependencies between components. Is there a way to express dependencies between packages in spack?

Reply:
Sorry I thought MKL etc. were part of the Base Toolkit e.g. described [here](https://software.intel.com/content/www/us/en/develop/tools/oneapi.html?cid=sem&source=sa360&campid=2020_q4_iags_us_iagsoapi_iagsoapiee_awa_text-link_brand_bmm_cd_dpd-oneapi-dpc-compiler_O-265LD_google_div_oos_non-pbm&ad_group=brand_oneapi-dpc-compiler_awa&intel_term=%2Bintel+%2Bdpc%2B%2B+%2Bcompiler&sa360id=43700053515352913&gclid=Cj0KCQjw59n8BRD2ARIsAAmgPmJCzCTB4pseogulRy3xUqmUJevLANZ0HgLrblhrzEGT2utpVF7KzIQaAtt_EALw_wcB) and the HPC Toolkit (which provided `icx` etc.) did not.

Regarding the size: the install as of now appears to occupy 4.4 GB, so I don't consider it prohibitively expensive to add DAL/MKL by default.

Reply:
You are correct, MKL is part of base toolkit. Download size of basekit is 2 GB, hpckit is 1 GB. Installed size of basekit is 10+ GB.

My thought was that IntelOneapi would be a base class with the code that:
* knows how to invoke the installer
* has the URL's for basekit/hpckit. 
* Has a dict that maps packages, e.g. mkl to the component name/installer that contains it.  
It would never be used directly to install anything.

All the packages that install something:
* intel-oneapi-compilers
* intel-oneapi-mkl
* intel-oneapi-tbb
* ...
would inherit from IntelOneapi, and override the set of packages that are installed. They would not have code to invoke the installer, or have to know whether it was in basekit/hpckit, etc.

Or maybe we don't need to make the installer code a baseclass, but a python module with utilities that all the oneapi packages rely on.

Reply:
> There are some dependencies between components. Is there a way to express dependencies between packages in spack?

(Forgot to answer this) yes: I think the `depends_on` declaration would be useful here. I think you are saying some of these components need others to work (and `depends_on` is intended for that).

The strategy suggested in https://github.com/spack/spack/pull/19506#discussion_r512677229 could work well (I think it would be implemented as a class in `build_systems` like the `IntelPackage` although not necessarily as complex).

IIRC packages like `intel-oneapi-mkl` will have open-source analogs. Do you know what the versioning scheme of these packages will look like (e.g. will they be in lockstep with the HPC Basekit versioning)? Is there a document which outlines the versioning scheme?

Reply:
There are 3 versions to consider. Open source tbb, product tbb (in basekit), and basekit (the collection of components). The 2 tbb's will have lockstep versioning, but the basekit version may not match the tbb version.

Unfortunately, there is no document that describes versioning.

Reply:
I think if we go this route we'll need a way to know what HPC toolkit version to download when we want a specific version of tbb. If it's not documented, is there a way to determine the supplied tbb version by examining the files after running an install?

Reply:
You can list the contents of a package without installing:
```
rscohn2@rscohn1-MOBL:l_HPCKit_b_2021.1.9.2205_offline$ ./install.sh --silent --list-components
Checking system requirements...
Done.
Wait while preparing installer...
Done.
Launching installer...
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-rscohn2'
ID                                      Version            Language Name
=========================================================================================================
intel.oneapi.lin.dpcpp-cpp-compiler-pro 2021.1-2198.beta09          Intel® oneAPI DPC++/C++ Compiler Pro
intel.oneapi.lin.ifort-compiler         2021.1-2198.beta09          Intel® Fortran Compiler Classic
intel.oneapi.lin.mpi.devel              2021.1-2099.beta09          Intel® MPI Library
intel.oneapi.lin.inspector              2021.1-319.beta09           Intel® Inspector
intel.oneapi.lin.clck                   2021.1-760.beta09           Intel® Cluster Checker
intel.oneapi.lin.itac                   2021.1-376.beta09           Intel® Trace Analyzer and Collector
rscohn2@rscohn1-MOBL:l_HPCKit_b_2021.1.9.2205_offline$
```
In addition to HPCkit, basekit, ,there will be standalone installers for each package (e.g. tbb, mkl). My motivation for recommending getting everything from hpckit and basekit was that it seemed more manageable to have less things to download. Maybe that does not matter so much. If you factor out the filename of the installer, the command lines are the same whether you use standalone or basekit.

Is this the procedure? Every time there is new release of basekit, someone downloads it and lists the contents, and then updates the packages that have new versions? Updating is just adding a new entry into a table with version, URL, and checksum.

Reply:
> My motivation for recommending getting everything from hpckit and basekit was that it seemed more manageable to have less things to download. Maybe that does not matter so much. If you factor out the filename of the installer, the command lines are the same whether you use standalone or basekit.

I don't think it matters a great deal to Spack: when making separate packages it is typically *easier* to have separate download locations for each one (I think both approaches are possible but FWIW I thought that was worth mentioning).

> Is this the procedure? Every time there is new release of basekit, someone downloads it and lists the contents, and then updates the packages that have new versions?

I think this is what the maintainers for the `oneapi` packages would have to do: the Spack packages need to list versions (and know where to download them from). This sounds workable though given the `install.sh` invocation you provide.

> In addition to HPCkit, basekit, ,there will be standalone installers for each package (e.g. tbb, mkl).

To make sure, could basekit be replaced entirely by calls to these standalone installers?

Reply:
> To make sure, could basekit be replaced entirely by calls to these standalone installers?

Yes, but I need to verify. 

> I don't think it matters a great deal to Spack: when making separate packages it is typically easier to have separate download locations for each one

Then it sounds best to use the standalones. Individual components may have out of cycle releases that would only be available via a standalone (not basekit).

Reply:
> Then it sounds best to use the standalones. Individual components may have out of cycle releases that would only be available via a standalone (not basekit).

Interesting: where would I download these individual packages?

Also, is this true of the components that come with the HPC toolkit? That is: is there a download for just the `icx`/`ifx` compiler?

Reply:
The standalones will be available in a few weeks. I have not seen them yet and do not know how they have decomposed the compilers.

Reply:
The standalones are available. https://software.intel.com/content/www/us/en/develop/articles/oneapi-standalone-components.html#dpcpp-cpp

How do you want to proceed? Unless you want to do it, I can take the file here and split it up into the base class for intel packages and then individual packages for each intel component.

Reply:
> Unless you want to do it, I can take the file here and split it up into the base class for intel packages and then individual packages for each intel component.

@rscohn2 sorry for the late response: that would be extremely useful IMO if you were to do it.

Does each component use this same script? For example from the linked site I see there is a download for MKL at https://registrationcenter-download.intel.com/akdlm/irc_nas/17402/l_onemkl_p_2021.1.1.52_offline.sh - I assume an MKL package could refer to this link directly without needing to use `./l_HPCKit_b_%s_offline.sh`). Or is the idea that generally speaking this and each component needs a similar set of environment variables and options and differs in terms of `--component`?

Also I'm assuming that `icx` and `icc` are still distributed together (although from your comment above `dpcpp` is available as a separate component and would not need to be included with the package that installs `icx`)

Reply:
> Also I'm assuming that icx and icc are still distributed together (although from your comment above dpcpp is available as a separate component and would not need to be included with the package that installs icx)

To clarify, if possible it would be good for future usage to separate out these compilers. It's fine if that won't work out, but if it's easy to do I think that will make compiler selection easier in the future. I got the impression that would not be easy for `icx`/`icc` but since `dpcpp` looked like a separate component I thought it might be possible in that case.

In more detail: fairly soon Spack will start treating compilers like other dependencies (e.g. libraries), so ideally it can retrieve the C compiler for a package with normal querying syntax like `root_spec['c']` and use that with the compiler wrapper; in that case the package would have an `executable` property which refers to the compiler exe; if the package contains multiple C compiler executables then this will be ambiguous. Also note that the way Spack currently models things, it won't be an issue for a single package to provide multiple compiler executables for the same language.